### PR TITLE
Add @if feature() syntax

### PR DESCRIPTION
**Pull request by peckpeck**
@if feature work like @if minimum_version but allow to distinguish
between features choosen at compile time.

### CODE REVIEW

**Comment by tzz** [comment]
I looked at this and overall it's good.  Some notes:

* Maybe there's a way to combine the two `@if` sections so code is not duplicated?  I am worried about the duplication of some tricky code.  Also we agreed not to nest `@if` on version, but now we need to agree on nesting for mixed `@if` macros.  My vote is to still disallow nesting but it may make life harder to library maintainers.

* The extra `feature.c` facility makes a lot of sense, but then `sysinfo.c:BuiltInClasses()` should be adjusted to walk through that feature list instead of duplicating the class names and logic.

**Comment by peckpeck** [comment]
The code looks similar but it's not duplicated. It's only the depth check that is duplicated.
I don't really think we need to nest @if macro.

I will update the PR with the static class creation.

**Comment by peckpeck** [comment]
@tzz PR updated

**Comment by tzz** [comment]
You should target master, not 3.7.x.  The developers do backporting.

File: libpromises/feature.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    if(r==0) {
+      return 1;
+    }
+  }
+  return 0;
+}
+
+void CreateHardClassesFromfeatures(EvalContext *ctx, char *tags)
+{
+  char vbuff[CF_BUFSIZE];
```
Comment (tzz) [inline libpromises/feature.c]: Please use a `Buffer` here, it's just as easy and doesn't have a size limit.

**Comment by tzz** [comment]
Looks OK to me, passing on to the developers...

**Comment by tzz** [comment]
The `@if` changes should get acceptance tests.

I would prefer if this PR focused on the feature piece and then you make a new PR to add the patch level to the version `@if`.  It makes review easier, but it's just my suggestion.  The developers can review them both here.  Either way, for that second change, you need to update the existing `@if` acceptance tests that currently only check major and minor.  It's not trivial because you have to test failure modes, not just success.  Sorry I have to make your life difficult :)

**Comment by peckpeck** [comment]
PR on master with tests: #2340

**Comment by kacfengine** [comment]
Closing in favor of #2340.

**Pull request closed without merging**